presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/StaticHiveCluster.java

```diff
@@ -65,13 +66,27 @@ public StaticHiveCluster(List<URI> metastoreUris, String metastoreUsername, Hive
     public HiveMetastoreClient createMetastoreClient()
             throws TException
     {
-        List<HostAndPort> metastores = new ArrayList<>(addresses);
-        Collections.shuffle(metastores.subList(1, metastores.size()));
-
         TException lastException = null;
-        for (HostAndPort metastore : metastores) {
+
+        for (HostAndPort metastore : addresses) {
             try {
-                HiveMetastoreClient client = clientFactory.create(metastore);
+                HiveMetastoreClient client = new FailureAwareHiveMetaStoreClient(clientFactory.create(metastore), new ResponseHandle()
+                {
+                    @Override
+                    public void success()
+                    {
+                        // do nothing
+                    }
+
+                    @Override
+                    public void failed(TException e)
+                    {
+                        if (!metastore.equals(addresses.getLast())) {
+                            addresses.remove(metastore);
```

Comment:
there should be some limit how many retries are performed before actually failing request. Otherwise we will loop infinitely.

Reply:
`retry` is done at `ThriftHiveMetastore`